### PR TITLE
HDDS-1888. Add containers to node2container map in SCM as part of ICR processing.

### PR DESCRIPTION
**Pull request by nandakumar131**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 46 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 589 | trunk passed |
| +1 | compile | 367 | trunk passed |
| +1 | checkstyle | 59 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 826 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 147 | trunk passed |
| 0 | spotbugs | 410 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 600 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 559 | the patch passed |
| +1 | compile | 368 | the patch passed |
| +1 | javac | 368 | the patch passed |
| +1 | checkstyle | 66 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 637 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 159 | the patch passed |
| +1 | findbugs | 616 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 285 | hadoop-hdds in the patch failed. |
| -1 | unit | 2232 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 42 | The patch does not generate ASF License warnings. |
| | | 7765 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.pipeline.TestNodeFailure |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis |
|   | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1211 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux d816cb2c6b10 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a7371a7 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/1/testReport/ |
| Max. process+thread count | 4259 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 598 | trunk passed |
| +1 | compile | 350 | trunk passed |
| +1 | checkstyle | 63 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 799 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 155 | trunk passed |
| 0 | spotbugs | 439 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 631 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 549 | the patch passed |
| +1 | compile | 349 | the patch passed |
| +1 | javac | 349 | the patch passed |
| +1 | checkstyle | 69 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 625 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 151 | the patch passed |
| +1 | findbugs | 642 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 288 | hadoop-hdds in the patch passed. |
| -1 | unit | 1600 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 47 | The patch does not generate ASF License warnings. |
| | | 7116 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.transport.server.ratis.TestCSMMetrics |
|   | hadoop.ozone.client.rpc.TestCloseContainerHandlingByClient |
|   | hadoop.hdds.scm.pipeline.TestNodeFailure |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |
|   | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.hdds.scm.pipeline.TestNode2PipelineMap |
|   | hadoop.ozone.client.rpc.TestBCSID |
|   | hadoop.ozone.om.TestScmSafeMode |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1211 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 4dc4d76ccc56 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e111789 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/2/testReport/ |
| Max. process+thread count | 3092 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/MockNodeManager.java [hunk 1]
```diff
@@ -267,6 +267,19 @@ public void removePipeline(Pipeline pipeline) {
...
+  @Override
+  public void addContainer(DatanodeDetails dd,
+                           ContainerID containerId)
+      throws NodeNotFoundException {
+    try {
+      Set<ContainerID> set = node2ContainerMap.getContainers(dd.getUuid());
+      set.add(containerId);
+      node2ContainerMap.setContainersForDatanode(dd.getUuid(), set);
+    } catch (SCMException e) {
+      e.printStackTrace();
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/MockNodeManager.java]: Is this needed ? can we just throw the exception or log the error.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java [hunk 2]
```diff
@@ -224,6 +224,23 @@ public NodeState getNodeState(UUID uuid) throws NodeNotFoundException {
...
+   * @throws NodeNotFoundException - if datanode is not known. For new datanode
+   *                        use addDatanodeInContainerMap call.
+   */
+  public void addContainer(final UUID uuid,
+                           final ContainerID containerId)
+      throws NodeNotFoundException {
+    if (!nodeToContainer.containsKey(uuid)) {
+      throw new NodeNotFoundException("Node UUID: " + uuid);
+    }
+    nodeToContainer.get(uuid).add(containerId);
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java]: the check and get calls can race with the remove call.

Also looking into the code, there is no remove call, we should remove the dn once it is dead ?

Comment (nandakumar131) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java]: For now we don't ever remove a datanode from NodeManager once it is registered. We should not remove dn once it is dead, we will not be able to show list of dead nodes to the user.
Maybe as part of decommissioning once the decommission is successful we can remove the dn.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/MockNodeManager.java (hunk 1 above)
Comment (nandakumar131) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/MockNodeManager.java]: We can only throw `NodeNotFoundException` or any runtime exception.
> log the error.
`printStackTrace` was added to log the error :) We don't have any logger defined in MockNodeManager.

The test-case using `MockNodeManager` will anyways depend on the existence of container after calling addContainer, and the test should eventually fail if this call fails.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 48 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 607 | trunk passed |
| +1 | compile | 370 | trunk passed |
| +1 | checkstyle | 76 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 851 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 162 | trunk passed |
| 0 | spotbugs | 421 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 616 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 564 | the patch passed |
| +1 | compile | 372 | the patch passed |
| +1 | javac | 372 | the patch passed |
| +1 | checkstyle | 70 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 656 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 166 | the patch passed |
| +1 | findbugs | 631 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 303 | hadoop-hdds in the patch passed. |
| -1 | unit | 2105 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 53 | The patch does not generate ASF License warnings. |
| | | 7806 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1211 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 952133fae2bc 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 70f4674 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/3/testReport/ |
| Max. process+thread count | 5009 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 92 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 696 | trunk passed |
| +1 | compile | 370 | trunk passed |
| +1 | checkstyle | 71 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 967 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 181 | trunk passed |
| 0 | spotbugs | 454 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 688 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 576 | the patch passed |
| +1 | compile | 440 | the patch passed |
| +1 | javac | 440 | the patch passed |
| +1 | checkstyle | 82 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 756 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 166 | the patch passed |
| +1 | findbugs | 699 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 358 | hadoop-hdds in the patch passed. |
| -1 | unit | 2206 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 41 | The patch does not generate ASF License warnings. |
| | | 8543 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.Test2WayCommitInRatis |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1211 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 757e082820c5 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3cc0ace |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/4/testReport/ |
| Max. process+thread count | 5278 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 90 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 620 | trunk passed |
| +1 | compile | 381 | trunk passed |
| +1 | checkstyle | 74 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 989 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 183 | trunk passed |
| 0 | spotbugs | 453 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 677 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 588 | the patch passed |
| +1 | compile | 428 | the patch passed |
| +1 | javac | 428 | the patch passed |
| +1 | checkstyle | 79 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 768 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 171 | the patch passed |
| +1 | findbugs | 724 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 374 | hadoop-hdds in the patch passed. |
| -1 | unit | 2202 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 43 | The patch does not generate ASF License warnings. |
| | | 8552 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.om.TestKeyManagerImpl |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1211 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux ac61aeb92bcc 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 00b5a27 |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/5/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/5/testReport/ |
| Max. process+thread count | 4547 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1211/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request closed without merging**